### ISSUE TITLE
Standard TemplateVMs should use the same default desktop software selection as the upstream OS

### ISSUE THREAD

**Issue by mfc** [QubesOS/qubes-issues#5367]
**The problem you're addressing (if any)**

We continue to run into cases where non-minimal templates lack basic user functionality (PDF reader, image viewer) which is **included in the desktop installation of the guest OS itself** but is not in the Qubes guest OS template.

For example, default desktop Gnome on Debian 10 comes with [`gnome-core`](https://packages.debian.org/buster/gnome-core), which includes basic applications like `evince`, `eog`... these are not included in the current Qubes Debian 10 template. 

I would like it to be clear what are we basing our OS templates on. When we say ["We try to respect each distro’s culture"](https://www.qubes-os.org/faq/#what-is-qubes-attitude-toward-changing-guest-distros), what are we choosing from that distro to present as the default?

for Debian, is it:
* [`task-desktop`](https://packages.debian.org/buster/task-desktop)?
* [`task-gnome-desktop`](https://packages.debian.org/buster/task-gnome-desktop)?
* something else

**Describe the solution you'd like**

For each guest OS template, clear documentation on what we base each guest OS template on.

**Where is the value to a user, and who might that user be?**

For non-minimal templates, providing the same initial functionality to the user as they expect in the OS the template is based on is [our policy](https://www.qubes-os.org/faq/#what-is-qubes-attitude-toward-changing-guest-distros).

**Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted**

https://www.qubes-os.org/faq/#what-is-qubes-attitude-toward-changing-guest-distros

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**

this issue is me again trying to approach the same topic of  https://github.com/QubesOS/qubes-issues/issues/1781 from another direction, as it continues to be a usability problem with Qubes.

**Comment by marmarek** [comment]
Right now we base on ["standard" task](https://wiki.debian.org/tasksel#A.22standard.22_task). It may be indeed a good idea to switch to `desktop` task, as indeed `standard` is minimal.
One challenge to overcome is installation media size. We're on the edge of DVD size. There are few options:
 - abandon the goal of fitting on DVD
 - reduce number of templates on the DVD (currently: Fedora, Debian, 2xWhonix)
 - split into two images: both above points

**Comment by mfc** [comment]
> Right now we base on ["standard" task](https://wiki.debian.org/tasksel#A.22standard.22_task). It may be indeed a good idea to switch to `desktop` task, as indeed `standard` is minimal.

okay great that is good to know. yeah agreed that I think our target should be `desktop`.

> One challenge to overcome is installation media size.

my preference (and the easiest solution) is to abandon goal of fitting on DVD. I would be interested in others' thoughts. 

should we email `qubes-users` and ask who uses DVDs to install? I would imagine some very small but vocal minority. but maybe they can use write-protected USB instead (if they use DVD for some security purpose). maybe that small minority could maintain a very minimal image with just one initial template? 

I would also be interested in understanding if it would be possible for Whonix not to ship as images, but instead could be built by the system based on the Debian 10 template if selected by user. probably would take a while. or if there doesn't need to be Whonix images at all, but instead just the latest Debian and some salt stack to configure the qubes to be used as Whonix qubes in the proper way. definitely sounds like something for further down the road tho.

**Comment by marmarek** [comment]
> I would also be interested in understanding if it would be possible for Whonix not to ship as images, but instead could be built by the system based on the Debian 10 template if selected by user. probably would take a while. or if there doesn't need to be Whonix images at all, but instead just the latest Debian and some salt stack to configure the qubes to be used as Whonix qubes in the proper way. definitely sounds like something for further down the road tho.

The main reason for including Whonix, is to allow single-file download bootstrapping Tor communication. It's much easier to download just one file (or maybe get it sent some other means) in restricted environment, than connecting your system to clearnet first to bootstrap other parts online. Main concern here is triggering some red flags while bootstrapping Tor/Whonix over clearnet.

Building Whonix on top of Debian locally would be doable, as as you said, it would take significant amount of time. Setting up automation for that also would take some time (both actual installation, and preparing installation packages set to be included on DVD).

**Comment by mfc** [comment]
good point, the bootstrap argument is very strong

**Comment by andrewdavidwong** [comment]
> the bootstrap argument is very strong

Agreed.

> split into two images

Providing different installers (a "minimal" version that fits on a DVD, a "full" version that includes everything, and maybe others in between) would be ideal from a user's perspective, but the time required to implement and maintain this might be better spent elsewhere.

I'm leaning toward "abandon the goal of fitting on DVD" for a few reasons:

* The capacity of a single-layer DVD remains fixed as technology progresses. It's not realistic to expect things to fit on DVDs forever, just as it wasn't realistic to expect things to fit on CDs forever. We migrated from CDs to DVDs. Why not from DVDs to Blu-rays? Sure, Blu-rays are more expensive, but new Qubes releases don't come out that often.
* There are also dual-layer DVDs.
* There are also USB drives with write-protect switches and firmware that is either non-rewritable or cryptographically-signed.

Related: https://www.qubes-os.org/doc/install-security/#choosing-an-installation-medium

This issue came up with the 3.2.1 ISO, and a few users weighed in here:
https://groups.google.com/d/topic/qubes-devel/PCkVcOqdIbc/discussion

**Comment by mfc** [comment]
> This issue came up with the 3.2.1 ISO, and a few users weighed in here:
> https://groups.google.com/d/topic/qubes-devel/PCkVcOqdIbc/discussion

yeah there is mentioned the argument of a minimal ISO that is just fedora and whonix (without debian), which makes more sense from a bootstrap perspective. I would say that is a good target if there is capacity/interest in a minimal ISO.

> I'm leaning toward "abandon the goal of fitting on DVD" for a few reasons:
> 
> * The capacity of a single-layer DVD remains fixed as technology progresses. It's not realistic to expect things to fit on DVDs forever, just as it wasn't realistic to expect things to fit on CDs forever. We migrated from CDs to DVDs. Why not from DVDs to Blu-rays? Sure, Blu-rays are more expensive, but new Qubes releases don't come out that often.
> * There are also dual-layer DVDs.
> * There are also USB drives with write-protect switches and firmware that is either non-rewritable or cryptographically-signed.

that is a nice overview of reasons, and I think a strong argument to abandon "single-layer DVD" ISO goal -- we don't even need to say we are abandoning fitting on a DVD, just on a single-layer DVD.

**Comment by DemiMarie** [comment]
Another option is to package the minimal templates by default, but offer to fetch the more full-featured ones from the network during the install process.

**Comment by marmarek** [comment]
I prefer to keep the network disabled during installation, as sys-net and related isolation is not in place yet. It could be done after first boot, but we'll need separate GUI for that. GUI for installing extra templates would be useful anyway.

**Comment by adrelanos** [comment]
@mfc 
> I would also be interested in understanding if it would be possible for Whonix not to ship as images, but instead could be built by the system based on the Debian 10 template if selected by user. 

@marmarek
> The main reason for including Whonix, is to allow single-file download bootstrapping Tor communication. It's much easier to download just one file (or maybe get it sent some other means) in restricted environment, than connecting your system to clearnet first to bootstrap other parts online. Main concern here is triggering some red flags while bootstrapping Tor/Whonix over clearnet.
>
> Building Whonix on top of Debian locally would be doable, as as you said, it would take significant amount of time. Setting up automation for that also would take some time (both actual installation, and preparing installation packages set to be included on DVD).

@marmarek
> I prefer to keep the network disabled during installation,

Possible in theory even with network disabled.

Quote https://www.whonix.org/wiki/Dev/Installation_from_Repository

> An installation of Debian can be transformed into Whonix ™ or Kicksecure ™. Also known as proverbial "sudo apt-get install whonix". This is also called distro-morphing.

This was implemented and now used in the wild. Whonix reported to be running on POWER9 using distro-morphing. References:

* https://wiki.raptorcs.com/wiki/Whonix
* https://wiki.raptorcs.com/wiki/Main_Page
* https://raptorcs.com
* https://en.wikipedia.org/wiki/POWER9
* https://wiki.debian.org/PPC64

Instead of a remote, online repository the Qubes installer could use a local, offline repository with Whonix packages. But indeed. Setting up automation for that would be quite some work.

**Comment by marmarek** [comment]
> Instead of a remote, online repository the Qubes installer could use a local, offline repository with Whonix packages.

Is it fully true? How TorBrowser is downloaded?

**Comment by adrelanos** [comment]
Tor Browser will be haunting me until the end of days. That indeed is the only exception. Options:

- [A] Either skip installation of Tor Browser. [1] [2] Or,

- [B] Give up on the download, verify Tor Browser archive through a script approach. [3] [4] Put the binary archive [5] into a package. [6]

Don't let the Tor Browser download issue block any progress on this one. [B] seems doable. I could do that part if the approach in this ticket "create Whonix using local repository during Qubes installed" is being worked on.

----

Footnotes:

[1] /usr/bin/torbrowser (by Whonix developers) would prompt to install after Tor Browser start in VM.

[2] Tor Browser download during tb-updater package installation can fail open since that should not break the package manager during a system upgrade. It just fails closed during template build since we want to ensure it is really installed by default.

[3] tb-updater / update-torbrowser script by Whonix developers

[4] That's really problematic anyhow.

[5] And signature for easier review.

[6] I've been contemplating that for a while now anyhow.

**Comment by rubyFeedback** [comment]
On http://distrowatch.org/ there is a new release (I just saw it), and also the note that people should use USB devices. I would like to comment on this - of course you can ignore this as-is, no problem, but I wanted to give you my opinion.

I here have an "oldschool DVD", for e. g. 4.3 gig or whatever the size limit is for DVD -R or DVD +R, and an optical reader device. I typically burn linux DVDs and test them out that way.

I also have several external USB devices, ranging between 4GIG to 3 terabytes or so. The latter in particular for several snapshots of my local data, over time, since I also went through data corruption, having lost work etc...

I believe I could quite easily get a decidated USB device, and then e. g. copy QubesOS on that, and work with that. This is in theory not a big problem.

BUT!

I still prefer DVDs for that particular job. The reason is primarily because I can label the DVD, it is quite cheap, and I can then store it somewhere without having to worry about it again at a later time. If I need it again, I just take the DVD and boot it up.

Yes, I could also use USB but ... my mind much prefers this strange distinction, and I prefer DVDs here. It just seems simpler for me to deal with it (I buy the USB in bulk device... e. g. 50 DVDs cost about 14 euro or so, which isn't that much really).

Due to this, I would like to ask the Qubes team if it would ALSO be possible to offer a slightly reduced size for the image as an additional download, one that fits into a regular DVD -R. I don't know if that is too much work, if so then just ignore it - but if possible, I'd much prefer to sacrifice some applications yet still be able to use a DVD (I can install stuff at a later time anyway via an internet connection or through other means). For me personally this would be more convenient - but I understand the problem of time constraints and people don't have infinite time, so please only consider it if it were simple to also add. (Does not have to happen at every release, could happen in yearly or bi-yearly release, such a DVD size; to me it also does not matter which programs would be skipped, but I would recommend skipping one or two of the larger ones).

> However, if you would prefer to use optical media, we suggest selecting a dual-layer DVD or a
> Blu-ray disc.

I can understand this too, but I do not use dual-layer DVD nor Blu-ray, and it is unlikely that I will. The cost-ratio benefit is not worth it for me to buy these, so I will stay with simple single-DVDs, most likely for many more years to come. I admit that it is odd because I also use USB devices a lot - but not exclusively. I also quickly abandoned CDs when DVDs came out, but I don't feel that switching away from single DVDs is as important as switching away from CDs, since the latter had that ~size limit of 700 mb which really was too little. Thank you for reading anyway!

**Comment by ninavizz** [comment]
I changed the title of this Issue, as this feels like much more of a service experience problem than a coding/process problem. 

Service experience problems, being how we communicate with users: in UI, and in docs. 

If I am incorrect in this assessment, pls to let me know, as I'm still coming up to speed on dumbing things down in my brain so all this techno _(insert gestural handwaving)_ stuff makes sense to me.

**Comment by andrewdavidwong** [comment]
> I changed the title of this Issue, as this feels like much more of a service experience problem than a coding/process problem.
> 
> Service experience problems, being how we communicate with users: in UI, and in docs.
> 

No, this issue is about changing the software selection to fit users' expectations, not the other way around. Even if we were to change users' expectations to fit the absence of software they need, they would still need it. If that software is already present by default in the standard "desktop" version of the upstream OS, then the same selection should be present in Qubes TemplateVMs running the same OS.

Once the appropriate software loadout (or "task" as Debian calls it) is selected, then merely inheriting this package selection from upstream should mean *less* work for the Qubes developers, as they should not (and hence need not) mess with the default software packages. I hope it works out this way in practice.

Of course, we will also document this, but it should be around one sentence.

None of this should have any effect on the minimal templates, which should remain minimal.

**Comment by andrewdavidwong** [comment]
> Right now we base on ["standard" task](https://wiki.debian.org/tasksel#A.22standard.22_task). It may be indeed a good idea to switch to `desktop` task, as indeed `standard` is minimal.

Sounds like this is all we need to do for Debian templates. Anything required for Fedora, @mfc, @marmarek?

**Comment by ben-grande** [comment]
@mfc Is this still the case?

**Comment by unman** [comment]
I have always found this an odd issue, as if there is ONE configuration
of Debian that a user might have.
As we provide different flavors of template, and the fuller templates do
provide some configuration of packages that include PDF,document,and
image viewer, the initial issue is resolved for Debian.
Cant speak for Fedora, but I suspect it is done there also.

**Comment by ben-grande** [comment]
> that include PDF,document,and image viewer, the initial issue is resolved for Debian. Cant speak for Fedora, but I suspect it is done there also.

It is fixed on Fedora. The image viewer is Ristretto and the PDF viewer is Atril. Reasonable options and easy to use.
